pkg/services/ngalert/notifier/channels/util.go

```diff
@@ -46,57 +46,51 @@ var (
 	ErrImagesUnavailable = errors.New("alert screenshots are unavailable")
 )
 
-// For each alert, attempts to load the models.Image for an image token
-// associated with the alert, then calls forEachFunc with the index of the
-// alert and the retrieved image struct. If there is no image token, or the
-// image does not exist, forEachFunc will be called with a nil value for the
-// image. If forEachFunc returns an error, withStoredImages will return
-// immediately. If there is a runtime error retrieving images from the image
-// store, withStoredImages will attempt to continue executing, after logging
-// a warning.
-func withStoredImages(ctx context.Context, l log.Logger, imageStore ImageStore, forEachFunc func(index int, image *models.Image) error, alerts ...*types.Alert) error {
-	for i := range alerts {
-		err := withStoredImage(ctx, l, imageStore, forEachFunc, i, alerts...)
-		if err != nil {
-			// Stop iteration as forEachFunc has found the intended image or
-			// iterated the maximum number of images
-			if errors.Is(err, ErrImagesDone) {
-				return nil
-			}
-			return err
-		}
-	}
-	return nil
-}
+type forEachImageFunc func(index int, image models.Image) error
 
-func withStoredImage(ctx context.Context, l log.Logger, imageStore ImageStore, imageFunc func(index int, image *models.Image) error, index int, alerts ...*types.Alert) error {
-	imgToken := getTokenFromAnnotations(alerts[index].Annotations)
-	if len(imgToken) == 0 {
-		err := imageFunc(index, nil)
-		if err != nil {
-			return err
-		}
+// getImage returns the image for the alert or an error. It returns a nil
+// image if the alert does not have an image token or the image does not exist.
+func getImage(ctx context.Context, l log.Logger, imageStore ImageStore, alert types.Alert) (*models.Image, error) {
+	token := getTokenFromAnnotations(alert.Annotations)
+	if token == "" {
+		return nil, nil
 	}
 
-	timeoutCtx, cancel := context.WithTimeout(ctx, ImageStoreTimeout)
-	img, err := imageStore.GetImage(timeoutCtx, imgToken)
-	cancel()
+	ctx, cancelFunc := context.WithTimeout(ctx, ImageStoreTimeout)
+	defer cancelFunc()
 
+	img, err := imageStore.GetImage(ctx, token)
 	if errors.Is(err, models.ErrImageNotFound) || errors.Is(err, ErrImagesUnavailable) {
-		err := imageFunc(index, nil)
-		if err != nil {
-			return err
-		}
+		return nil, nil
 	} else if err != nil {
-		// Ignore errors. Don't log "ImageUnavailable", which means the storage doesn't exist.
-		l.Warn("failed to retrieve image url from store", "err", err)
+		l.Warn("failed to get image with token", "token", token, "err", err)
+		return nil, err
```

Comment:
Returning and `error` here is a deviation in functionality from the previous version. Just want to make sure this is intentional and desired.

Previously, these "unexpected" errors would call the `imageFunc` with a `nil` image and would be ignored by the receivers. Iteration in `withStoredImages` would continue to the next alert.

Now, these "unexpected" errors will cause `withStoredImages` to return early with an error, no longer continuing to the next alert.

```
func withStoredImages
...
		img, err := getImage(ctx, l, imageStore, *alert)
		if err != nil {
			return err
...
```

This could cause a single intermittent db error to stop all images being attached to a notification.

Reply:
Thanks for the review Matt! That is intentional for now as I am working on the suspicion that unless the row contains invalid data that cannot be deserialized, errors between the client and the database tend to happen in bursts (i.e a short period of packet loss), a partition, or something else that won't have resolved before the next alert is iterated, but should resolve before the next evaluation or `repeat_interval`.

The thought here is something like "oh an unexpected error occurred getting the first image. The chance of the same error occurring for the second image, third image, etc, is quite high because something is wrong (either ephemeral or persistent), let's just get the notification out".

I'm 100% open to discussion on this though!